compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/ConstantValueGenerator.kt

```diff
@@ -36,23 +36,32 @@ abstract class ConstantValueGenerator(
         startOffset: Int,
         endOffset: Int,
         constantValue: ConstantValue<*>,
-        varargElementType: KotlinType? = null
     ): IrExpression =
         // Assertion is safe here because annotation calls and class literals are not allowed in constant initializers
-        generateConstantOrAnnotationValueAsExpression(startOffset, endOffset, constantValue, null, varargElementType)!!
+        generateConstantOrAnnotationValueAsExpression(startOffset, endOffset, constantValue, null, null)!!
 
     /**
      * @return null if the constant value is an unresolved annotation or an unresolved class literal
      */
+    fun generateAnnotationValueAsExpression(
+        startOffset: Int,
+        endOffset: Int,
+        constantValue: ConstantValue<*>,
+        valueParameter: ValueParameterDescriptor,
+    ): IrExpression? =
+        generateConstantOrAnnotationValueAsExpression(
+            startOffset, endOffset, constantValue, valueParameter.type, valueParameter.varargElementType
+        )
+
     private fun generateConstantOrAnnotationValueAsExpression(
         startOffset: Int,
         endOffset: Int,
         constantValue: ConstantValue<*>,
-        realType: KotlinType?,
-        varargElementType: KotlinType? = null
+        inferredType: KotlinType?,
+        inferredArrayElementType: KotlinType?
     ): IrExpression? {
         val constantValueType = constantValue.getType(moduleDescriptor)
-        val constantKtType = realType ?: constantValueType
+        val constantKtType = inferredType ?: constantValueType
```

Comment:
By the way, the fact that `constantValueType` is used in two places instead of `constantKtType` hints at incorrect generation for annotations too:

 * in spreadOperatorInAnnotationArguments, `@A(*arrayOf("a"), *arrayOf("b"))` is translated into `A(xs = [['a'], ['b']])` instead of `A(xs = ['a', 'b'])` - this is an IrText test even, so you can see that in the test data;
 * in annotationWithKotlinProperty, `@Foo(KotlinClass.FOO_INT)` is parsed as if `KotlinClass.FOO_INT` is an enum value - except it's really a `const val` with a type of `Int`, so we just silently fail to translate the argument.

Reply:
Let's either 1) fix this, 2) add TODOs, or 3) report to YouTrack

Reply:
Added two TODOs.